### PR TITLE
router: add envoy-ratelimited retry policy

### PR DESCRIPTION
**Pull request by numerodix**
This PR implements a new retry policy that retries a request if the response contains the header `x-envoy-ratelimited`.

This is PR (1/2) that addresses issue #12200. See also the design document: https://docs.google.com/document/d/1NSzrx3-KsAFs0ObaLQZUVIt9O4PZv3mXi3Lm7LNzmZE/edit?usp=sharing

Signed-off-by: Martin Matusiak <numerodix@gmail.com>

Risk Level: Low
Testing: New unit test added.
Docs Changes: added
Release Notes: added

### CODE REVIEW

File: source/common/router/retry_state_impl.cc [hunk 1]
```diff
@@ -290,9 +293,9 @@ RetryStatus RetryStateImpl::shouldHedgeRetryPerTryTimeout(DoRetryCallback callba
 }
 
 bool RetryStateImpl::wouldRetryFromHeaders(const Http::ResponseHeaderMap& response_headers) {
-  // We never retry if the request is rate limited.
+  // We retry our own rate limited requests only when the envoy-ratelimited policy is in effect.
```
Comment (snowp) [inline source/common/router/retry_state_impl.cc:296]: "Our own" implies to me that this envoy instance rate limited the request, but its likely going to be another Envoy instance

Comment (numerodix) [reply source/common/router/retry_state_impl.cc:296]: Yes, that's a good point. I'll rephrase this to make it clear we're talking about an upstream envoy.
